File: app/src/main/java/com/rakangsoftware/recyclerviewdemo/MainActivity.java
package com.rakangsoftware.recyclerviewdemo;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Get a reference to the RecyclerView
        RecyclerView recyclerView = findViewById(R.id.list);

        // Create and set the layout manager
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);

        recyclerView.setLayoutManager(layoutManager);

        // Set the adaptor
        CountryAdapter adapter = new CountryAdapter();
        recyclerView.setAdapter(adapter);

        // Load data into the adapter.
        CountryRepository countryRepository = new CountryRepository();
        adapter.setCountryList(countryRepository.getAll());

        adapter.setOnCountryClickListener(new OnCountryClickListener() {
            @Override
            public void onCountryClicked(final Country country) {
                Log.d(TAG, "Country: " + country.getCountryName() + " clicked!");
            }
        });
    }
}

File: app/src/main/java/com/rakangsoftware/recyclerviewdemo/MainActivity.kt
package com.rakangsoftware.recyclerviewdemo

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Create and set the layout manager
        val layoutManager = LinearLayoutManager(this)
        layoutManager.orientation = LinearLayoutManager.VERTICAL

        // Set the adaptor
        val adapter = CountryAdapter()

        list.layoutManager = layoutManager
        list.adapter = adapter

        // Load data into the adapter.
        val countryRepository = CountryRepository()
        adapter.setCountryList(countryRepository.all)

        adapter.setOnCountryClickListener(object : OnCountryClickListener {
            override fun onCountryClicked(country: Country) {
                Log.d(TAG, "Country: ${country.countryName} clicked!")
            }
        })
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}
